# Trực quan hóa Dữ liệu (Data Visualization)

Thời lượng ước tính: **30** phút

## Mục tiêu

Sau khi hoàn thành lab, này bạn sẽ có thể:

*   Tạo trực quan hóa dữ liệu với Python
*   Sử dụng các thư viện Python khác nhau cho trực quan hóa


## Giới thiệu

Các lab này sẽ giới thiệu cho bạn trực quan hóa dữ liệu với Python cách cụ thể và nhất quán nhất có thể.
Về tính nhất quán, vì không có thư viện trực quan hóa dữ liệu nào *tốt nhất* khả dụng cho Python - cho đến khi tạo các lab này - chúng ta cần giới thiệu các thư viện khác nhau và cho thấy lợi ích của chúng khi thảo luận về các khái niệm trực quan hóa mới. Bằng cách này, chúng tôi hy vọng sẽ giúp học viên hiểu rõ về các thư viện và khái niệm trực quan để họ có thể đánh giá và quyết định kỹ thuật và công cụ trực quan hóa tốt nhất cho một bài toán *và* đối tượng nhất định.
Hãy đảm bảo bạn đã hoàn thành các điều kiện tiên quyết cho khóa học này, cụ thể là [**Python cơ bản cho khoa học dữ liệu**](https://www.edx.org/course/python-basics-for-data-science-2?utm_medium=Exinfluencer\&utm_source=Exinfluencer\&utm_content=000026UJ\&utm_term=10006555\&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ENSkillsNetwork20297740-2021-01-01) và [**Phân tích dữ liệu với Python**](https://www.edx.org/course/data-analysis-with-python?utm_medium=Exinfluencer\&utm_source=Exinfluencer\&utm_content=000026UJ\&utm_term=10006555\&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ENSkillsNetwork20297740-2021-01-01).

**Lưu ý**: Phần lớn các đồ thị và hình ảnh trực quan sẽ được tạo bằng cách sử dụng dữ liệu được lưu trữ trong dataframe *pandas*. Do đó, trong lab này, chúng tôi cung cấp một khóa học sơ lược về *pandas*. Tuy nhiên, nếu bạn muốn tìm hiểu thêm về thư viện *pandas* thì chúng tôi cũng cung cấp các mô tả và giải thích chi tiết về cách sử dụng cũng như cách dọn dẹp và xử lý dữ liệu được lưu trữ trong dataframe *pandas* [**Phân tích dữ liệu với Python**](https://www.edx.org/course/data-analysis-with-python?utm_medium=Exinfluencer\&utm_source=Exinfluencer\&utm_content=000026UJ\&utm_term=10006555\&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ENSkillsNetwork20297740-2021-01-01).

***


## Mục lục

<div class="alert alert-block alert-info" style="margin-top: 20px">

1.  [Khám phá tập dữ liệu với *pandas*](#0)<br>
    1.1 [Tập dữ liệu: Nhập cư vào Canada từ 1980 đến 2013](#2)<br>
    1.2 [*Pandas* cơ bản](#4) <br>
    1.3 [*Pandas* trung cấp: Lập chỉ mục và lựa chọn](#6) <br>
2.  [Trực quan hóa Dữ liệu sử dụng Matplotlib](#8) <br>
    2.1 [Matplotlib: Thư viện Trực quan hóa Python chuẩn](#10) <br>
3.  [Biểu đồ đường](#12)

</div>


# Khám phá tập dữ liệu với *pandas* <a id="0"></a>

*pandas* là bộ công cụ phân tích dữ liệu cần thiết cho Python. Từ [website](http://pandas.pydata.org/?utm_medium=Exinfluencer\&utm_source=Exinfluencer\&utm_content=000026UJ\&utm_term=10006555\&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ENSkillsNetwork20297740-2021-01-01):

> *pandas* là gói Python cung cấp cấu trúc dữ liệu nhanh, linh hoạt và thể hiện được thiết kế giúp làm việc với dữ liệu “quan hệ” hoặc “có nhãn” dễ dàng và trực quan. Nó nhằm mục đích trở thành khối xây dựng cấp cao căn bản để thực hiện phân tích dữ liệu **thực tiễn** trong Python.

Khóa học căn cứ vào _pandas_ để xử lý dữ liệu, phân tích và trực quan hóa. Chúng tôi khuyến khích bạn dành chút thời gian và tự làm quen với Tham chiếu API *pandas*: http://pandas.pydata.org/pandas-docs/stable/api.html.


## Tập dữ liệu: Nhập cư vào Canada từ 1980 đến 2013 <a id="2"></a>


Nguồn tập dữ liệu: [International migration flows to and from selected countries - The 2015 revision](http://www.un.org/en/development/desa/population/migration/data/empirical2/migrationflows.shtml?utm_medium=Exinfluencer\&utm_source=Exinfluencer\&utm_content=000026UJ\&utm_term=10006555\&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ENSkillsNetwork20297740-2021-01-01).

Tập dữ liệu chứa dữ liệu hàng năm về dòng người nhập cư quốc tế được ghi lại bởi các quốc gia điểm đến. Dữ liệu thể hiện cả dòng vào và dòng ra theo nơi sinh, quốc tịch hoặc nơi cư trú trước đây / tiếp theo của cả người nước ngoài và người mang quốc tịch. Phiên bản hiện tại hiển thị dữ liệu liên quan đến 45 quốc gia.

Trong lab này, chúng ta sẽ tập trung vào dữ liệu nhập cư Canada.

<img src = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/labs/Module%201/images/DataSnapshot.png" align="center" width=900>

Tập dữ liệu Nhập cư tới Canada có thể truy xuất <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/Data%20Files/Canada.xlsx?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ENSkillsNetwork20297740-2021-01-01">tại đây</a>.

***


## *Pandas* cơ bản<a id="4"></a>


Trước tiên, chúng ta sẽ import 2 mô-đun phân tích dữ liệu chính: *pandas* và **Numpy**.


In [26]:
import numpy as np  # hữu ích với nhiều tính toán khoa học trong Python
import pandas as pd # thư viện cấu trúc dữ liệu chính

Hãy download và import tập dữ liệu Canadian Immigration chính của chúng tôi bằng phương thức _pandas_ `read_excel()`. Thông thường, chúng ta sẽ cần download một mô-đun _pandas_ yêu cầu đọc trong file excel trước khi có thể làm điều đó. Mô-đun này là **xlrd**. Để thuận tiện cho bạn, chúng tôi đã cài đặt sẵn mô-đun này, vì vậy bạn sẽ không phải lo lắng về điều đó. Nếu không, bạn sẽ cần chạy dòng code sau để cài đặt mô-đun **xlrd**:
```
!conda install -c anaconda xlrd --yes
```


Bây giờ, chúng ta đã sẵn sàng để đọc dữ liệu.


In [27]:
df_can = pd.read_excel('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/Data%20Files/Canada.xlsx',
                        sheet_name='Canada by Citizenship',
                       skiprows=range(20),
                       skipfooter=2)

print ('Data read into a pandas dataframe!')

Data read into a pandas dataframe!


Hãy xem 5 hàng trên cùng của tập dữ liệu bằng hàm `head()`.

In [28]:
df_can.head()
# tip: Bạn có thể chỉ định số hàng bạn muốn như sau: df_can.head(10)

Type    Coverage          OdName  AREA AreaName   REG  \
0  Immigrants  Foreigners     Afghanistan   935     Asia  5501   
1  Immigrants  Foreigners         Albania   908   Europe   925   
2  Immigrants  Foreigners         Algeria   903   Africa   912   
3  Immigrants  Foreigners  American Samoa   909  Oceania   957   
4  Immigrants  Foreigners         Andorra   908   Europe   925   

           RegName  DEV             DevName  1980  ...  2004  2005  2006  \
0    Southern Asia  902  Developing regions    16  ...  2978  3436  3009   
1  Southern Europe  901   Developed regions     1  ...  1450  1223   856   
2  Northern Africa  902  Developing regions    80  ...  3616  3626  4807   
3        Polynesia  902  Developing regions     0  ...     0     0     1   
4  Southern Europe  901   Developed regions     0  ...     0     0     1   

   2007  2008  2009  2010  2011  2012  2013  
0  2652  2111  1746  1758  2203  2635  2004  
1   702   560   716   561   539   620   603  
2  3623  4005  5393  4752  4325  3774  4331  
3     0     0     0     0     0     0     0  
4     1     0     0     0     0     1     1  

[5 rows x 43 columns]

Chúng ta cũng có thể xem 5 hàng dưới cùng của tập dữ liệu bằng hàm `tail()`.

In [29]:
df_can.tail()

Type    Coverage          OdName  AREA AreaName  REG  \
190  Immigrants  Foreigners        Viet Nam   935     Asia  920   
191  Immigrants  Foreigners  Western Sahara   903   Africa  912   
192  Immigrants  Foreigners           Yemen   935     Asia  922   
193  Immigrants  Foreigners          Zambia   903   Africa  910   
194  Immigrants  Foreigners        Zimbabwe   903   Africa  910   

                RegName  DEV             DevName  1980  ...  2004  2005  2006  \
190  South-Eastern Asia  902  Developing regions  1191  ...  1816  1852  3153   
191     Northern Africa  902  Developing regions     0  ...     0     0     1   
192        Western Asia  902  Developing regions     1  ...   124   161   140   
193      Eastern Africa  902  Developing regions    11  ...    56    91    77   
194      Eastern Africa  902  Developing regions    72  ...  1450   615   454   

     2007  2008  2009  2010  2011  2012  2013  
190  2574  1784  2171  1942  1723  1731  2112  
191     0     0     0     0     0     0     0  
192   122   133   128   211   160   174   217  
193    71    64    60   102    69    46    59  
194   663   611   508   494   434   437   407  

[5 rows x 43 columns]

Khi phân tích một tập dữ liệu, bạn nên bắt đầu bằng cách lấy thông tin cơ bản về dataframe. Chúng ta có thể làm điều này bằng phương thức `info()`.

Có thể sử dụng phương thức này để có được một bản tóm tắt ngắn gọn về dataframe.

In [30]:
df_can.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Columns: 43 entries, Type to 2013
dtypes: int64(37), object(6)
memory usage: 65.6+ KB


Để có được list các tiêu đề cột, chúng ta có thể gọi tham số `.columns` của dataframe.




In [31]:
df_can.columns.values

array(['Type', 'Coverage', 'OdName', 'AREA', 'AreaName', 'REG', 'RegName',
       'DEV', 'DevName', 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987,
       1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998,
       1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009,
       2010, 2011, 2012, 2013], dtype=object)

Tương tự, để lấy list các chỉ mục, chúng ta sử dụng tham số `.index`.

In [32]:
df_can.index.values

array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,
        13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,
        26,  27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,
        39,  40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,
        52,  53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,
        65,  66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,
        78,  79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,
        91,  92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103,
       104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116,
       117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129,
       130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142,
       143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155,
       156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168,
       169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 18

Lưu ý: Kiểu index và columns mặc định KHÔNG phải là list.

In [33]:
print(type(df_can.columns))
print(type(df_can.index))

<class 'pandas.core.indexes.base.Index'>
<class 'pandas.core.indexes.range.RangeIndex'>


Để lấy chỉ mục và các cột dưới dạng list, chúng ta có thể sử dụng phương thức `tolist()`.

In [34]:
df_can.columns.tolist()
df_can.index.tolist()

print (type(df_can.columns.tolist()))
print (type(df_can.index.tolist()))

<class 'list'>
<class 'list'>


Để xem các kích thước của dataframe, chúng ta sử dụng tham số `.shape`.

In [35]:
# kích thước của dataframe (hàng, cột)
df_can.shape

(195, 43)

Lưu ý: Các kiểu chính được lưu trữ trong đối tượng _pandas_ là _float_, _int_, _bool_, _datetime64[ns]_ và  _datetime64[ns, tz] (in >= 0.17.0)_, _timedelta[ns]_, _category (in >= 0.15.0)_, và _object_ (string). Ngoài ra, các kiểu này có kích thước khoản mục (item size), chằng hạn: int64 và int32.

Hãy làm sạch tập dữ liệu để loại bỏ một vài cột không cần thiết. Chúng ta có thể sử dụng phương thức pandas `drop()` như sau:

In [36]:
# trong pandas axis=0 biểu diễn hàng (mặc định) và axis=1 biểu diễn cột.
df_can.drop(['AREA','REG','DEV','Type','Coverage'], axis=1, inplace=True)
df_can.head(2)

OdName AreaName          RegName             DevName  1980  1981  \
0  Afghanistan     Asia    Southern Asia  Developing regions    16    39   
1      Albania   Europe  Southern Europe   Developed regions     1     0   

   1982  1983  1984  1985  ...  2004  2005  2006  2007  2008  2009  2010  \
0    39    47    71   340  ...  2978  3436  3009  2652  2111  1746  1758   
1     0     0     0     0  ...  1450  1223   856   702   560   716   561   

   2011  2012  2013  
0  2203  2635  2004  
1   539   620   603  

[2 rows x 38 columns]

Hãy đổi tên các cột để chúng có ý nghĩa. Chúng ta có thể sử dụng phương thức `rename ()` bằng cách chuyển vào dictionary các tên cũ và mới như sau:

In [37]:
df_can.rename(columns={'OdName':'Country', 'AreaName':'Continent', 'RegName':'Region'}, inplace=True)
df_can.columns

Index([  'Country', 'Continent',    'Region',   'DevName',        1980,
              1981,        1982,        1983,        1984,        1985,
              1986,        1987,        1988,        1989,        1990,
              1991,        1992,        1993,        1994,        1995,
              1996,        1997,        1998,        1999,        2000,
              2001,        2002,        2003,        2004,        2005,
              2006,        2007,        2008,        2009,        2010,
              2011,        2012,        2013],
      dtype='object')

Chúng ta cũng sẽ thêm cột 'Total' tính tổng số người nhập cư theo quốc gia trong toàn bộ giai đoạn 1980 - 2013, như sau:

In [38]:
df_can['Total'] = df_can.sum(axis=1)

TypeError: can only concatenate str (not "int") to str

Chúng ta có thể kiểm tra xem có bao nhiêu đối tượng null trong tập dữ liệu như sau:

In [ ]:
df_can.isnull().sum()

Country      0
Continent    0
Region       0
DevName      0
1980         0
1981         0
1982         0
1983         0
1984         0
1985         0
1986         0
1987         0
1988         0
1989         0
1990         0
1991         0
1992         0
1993         0
1994         0
1995         0
1996         0
1997         0
1998         0
1999         0
2000         0
2001         0
2002         0
2003         0
2004         0
2005         0
2006         0
2007         0
2008         0
2009         0
2010         0
2011         0
2012         0
2013         0
dtype: int64

Cuối cùng, hãy xem tóm tắt nhanh từng cột trong dataframe bằng phương thức `describe()`.

In [ ]:
df_can.describe()

***

## *Pandas* trung cấp: Lập chỉ mục và Lựa chọn<a id="6"></a>


### Chọn cột

**Có 2 cách để lọc tên cột:**

Phương pháp 1: Nhanh chóng và dễ dàng, nhưng chỉ hoạt động nếu tên cột KHÔNG CÓ khoảng cách hoặc ký tự đặc biệt.

```python
    df.column_name
        (returns series)
```

Phương pháp 2: Mạnh hơn và có thể lọc ở nhiều cột.

```python
    df['column']  
        (returns series)
```

```python
    df[['column 1', 'column 2']]
        (returns dataframe)
```

***


Ví dụ: Hãy thử lọc ở danh sách các quốc gia ('Country').


In [ ]:
df_can.Country  # trả về một chuỗi

0         Afghanistan
1             Albania
2             Algeria
3      American Samoa
4             Andorra
            ...      
190          Viet Nam
191    Western Sahara
192             Yemen
193            Zambia
194          Zimbabwe
Name: Country, Length: 195, dtype: object

Hãy thử lọc ở danh sách các quốc gia ('OdName') và dữ liệu cho các năm: 1980 - 1985.


In [ ]:
df_can[['Country', 1980, 1981, 1982, 1983, 1984, 1985]] # trả về một dataframe
# lưu ý rằng 'Country' là một string, và các năm là các số nguyên.
# về tính thống nhất, chúng ta sẽ chuyển tất cả các tên thành string sau.

Country  1980  1981  1982  1983  1984  1985
0       Afghanistan    16    39    39    47    71   340
1           Albania     1     0     0     0     0     0
2           Algeria    80    67    71    69    63    44
3    American Samoa     0     1     0     0     0     0
4           Andorra     0     0     0     0     0     0
..              ...   ...   ...   ...   ...   ...   ...
190        Viet Nam  1191  1829  2162  3404  7583  5907
191  Western Sahara     0     0     0     0     0     0
192           Yemen     1     2     1     6     0    18
193          Zambia    11    17    11     7    16     9
194        Zimbabwe    72   114   102    44    32    29

[195 rows x 7 columns]

### Chọn hàng

Có 3 cách chính để chọn hàng:

```python
    df.loc[label]        
        #lọc theo nhãn của chỉ mục/cột
    df.iloc[index]       
        #lọc theo vị trí của chỉ mục/cột
```


Trước khi tiếp tục, hãy lưu ý rằng chỉ mục mặc định của tập dữ liệu là một dãy số từ 0 đến 194. Điều này gây khó khăn cho việc thực hiện truy vấn theo một quốc gia cụ thể. Ví dụ để tìm kiếm dữ liệu về Nhật Bản, cần biết giá trị chỉ mục tương ứng.

Điều này có thể được khắc phục dễ dàng bằng cách đặt cột 'Country' làm chỉ mục bằng phương thức `set_index()`.

In [ ]:
df_can.set_index('Country', inplace=True)
# tip: Đối lập của set là reset. Vậy để reset index, chúng ta có thể dùng df_can.reset_index()

In [ ]:
df_can.head(3)

In [ ]:
# tùy chọn: để loại tên của index
df_can.index.name = None

Ví dụ: Hãy xem số lượng người nhập cư từ Nhật Bản (hàng 87) trong các tình huống sau:
1\. Dữ liệu hàng đầy đủ (tất cả các cột)
2\. Trong năm 2013
3\. Trong những năm 1980 đến 1985


In [ ]:
# 1. dữ liệu hàng đầy đủ (tất cả các cột)
print(df_can.loc['Japan'])

# các phương thức thay thế
print(df_can.iloc[87])
print(df_can[df_can.index == 'Japan'].T.squeeze())

In [ ]:
# 2. trong năm 2013
print(df_can.loc['Japan', 2013])

# phương thức thay thế
print(df_can.iloc[87, 36]) # năm 2013 là cột cuối, với chỉ số vị trí là 36

In [ ]:
# 3. trong các năm từ 1980 tới 1985
print(df_can.loc['Japan', [1980, 1981, 1982, 1983, 1984, 1984]])
print(df_can.iloc[87, [3, 4, 5, 6, 7, 8]])

Tên cột là số nguyên (chẳng hạn như năm) có thể gây ra một số nhầm lẫn. Ví dụ: khi tham khảo năm 2013, ai đó có thể nhầm lẫn với chỉ số vị trí thứ 2013.

Để tránh sự bất thường này, hãy chuyển đổi tên cột thành string: '1980' to '2013'.

In [ ]:
df_can.columns = list(map(str, df_can.columns))
# [in (type(x)) cho x trong df_can.columns.values] #<-- bỏ chú thích để kiểm tra kiểu của tiêu đề cột

Do chúng ta đã chuyển đổi các năm thành string, hãy khai báo một biến để cho phép chúng ta dễ dàng gọi toàn bộ phạm vi các năm:

In [ ]:
# hữu ích cho vẽ biểu đồ sau này
years = list(map(str, range(1980, 2014)))
years

### Lọc theo tiêu chí

Để lọc dataframe dựa trên một điều kiện, chúng ta chỉ cần chuyển điều kiện thành một vectơ boolean.

Ví dụ: Hãy lọc dataframe để hiển thị dữ liệu về các quốc gia Châu Á (AreaName = Asia).

In [ ]:
# 1. tạo chuỗi boolean điều kiện
condition = df_can['Continent'] == 'Asia'
print(condition)

In [ ]:
# 2. truyền điều kiện này vào dataframe
df_can[condition]

In [ ]:
# chúng ta có thể truyền nhiều tiêu chí vào cùng một dòng
# hãy lọc cho AreaNAme = Asia và RegName = Southern Asia

df_can[(df_can['Continent']=='Asia') & (df_can['Region']=='Southern Asia')]

# lưu ý: Khi sử dụng toán tử 'and' và 'or', pandas yêu cầu chúng ta dùng '&' và '|' thay vì 'and' và 'or'
# đừng quên đặt 2 điều kiện trong dấu ngoặc đơn

Trước khi tiến hành: hãy xem lại các thay đổi mà chúng ta đã tạo với dataframe.


In [ ]:
print('data dimensions:', df_can.shape)
print(df_can.columns)
df_can.head(2)

***

# Trực quan hóa dữ liệu với Matplotlib<a id="8"></a>


## Matplotlib: Thư viện trực quan hóa Python chuẩn<a id="10"></a>

[Matplotlib](http://matplotlib.org/?utm_medium=Exinfluencer\&utm_source=Exinfluencer\&utm_content=000026UJ\&utm_term=10006555\&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ENSkillsNetwork20297740-2021-01-01 ) là thư viện vẽ biểu đồ chính mà chúng ta sẽ khám phá trong khóa học.  Như đã đề cập trên website:

> Matplotlib là một thư viện vẽ biểu đồ Python 2D, tạo ra các hình chất lượng xuất bản ở nhiều định dạng bản cứng và môi trường tương tác trên các nền tảng. Matplotlib có thể được sử dụng trong các tập lệnh Python, Python và IPython shell, Jupyter notebook, máy chủ ứng dụng web và 4 bộ công cụ giao diện người dùng đồ họa.

Matplotlib là một công cụ cần thiết để tạo hình ảnh trực quan có tác động với python.


### Matplotlib.Pyplot

Một trong những khía cạnh cốt lõi của Matplotlib là `matplotlib.pyplot`. Đó là lớp scripting của Matplotlib mà chúng ta đã nghiên cứu chi tiết trong các video về Matplotlib. Hãy nhớ lại rằng nó là một tập hợp các hàm kiểu câu lệnh giúp cho Matplotlib hoạt động giống như MATLAB. Mỗi hàm `pyplot` thực hiện một số thay đổi đối với một hình: ví dụ: tạo một hình, tạo vùng vẽ đồ thị trong một hình, vẽ một số đường trong vùng vẽ đồ thị, trang trí đồ thị bằng nhãn,... Trong lab này, chúng ta sẽ làm việc với lớp scripting để tìm hiểu cách tạo các biểu đồ đường. Trong các lab sau, chúng ta cũng sẽ làm việc với lớp Artist để thử nghiệm trực tiếp xem nó khác với lớp scripting thế nào.

Hãy import `Matplotlib` và `Matplotlib.pyplot` như sau:


In [ ]:
# hãy sử dụng inline backend
%matplotlib inline

import matplotlib as mpl
import matplotlib.pyplot as plt

\*tùy chọn: kiểm tra xem Matplotlib đã load chưa.


In [ ]:
print ('Matplotlib version: ', mpl.__version__) # >= 2.0.0

Matplotlib version:  3.8.3


\*tùy chọn: áp dụng một style cho Matplotlib.


In [ ]:
print(plt.style.available)
mpl.style.use(['ggplot']) # tùy chọn: style ggplot

['Solarize_Light2', '_classic_test_patch', '_mpl-gallery', '_mpl-gallery-nogrid', 'bmh', 'classic', 'dark_background', 'fast', 'fivethirtyeight', 'ggplot', 'grayscale', 'seaborn-v0_8', 'seaborn-v0_8-bright', 'seaborn-v0_8-colorblind', 'seaborn-v0_8-dark', 'seaborn-v0_8-dark-palette', 'seaborn-v0_8-darkgrid', 'seaborn-v0_8-deep', 'seaborn-v0_8-muted', 'seaborn-v0_8-notebook', 'seaborn-v0_8-paper', 'seaborn-v0_8-pastel', 'seaborn-v0_8-poster', 'seaborn-v0_8-talk', 'seaborn-v0_8-ticks', 'seaborn-v0_8-white', 'seaborn-v0_8-whitegrid', 'tableau-colorblind10']


### Vẽ biểu đồ trong *pandas*

May thay, pandas có cài sẵn Matplotlib mà chúng ta có thể sử dụng. Vẽ đồ thị trong *pandas* đơn giản như việc thêm phương thức `.plot ()` vào một series hoặc dataframe.

Tài liệu:

*   [Vẽ biểu đồ với Series](http://pandas.pydata.org/pandas-docs/stable/api.html?utm_medium=Exinfluencer\&utm_source=Exinfluencer\&utm_content=000026UJ\&utm_term=10006555\&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ENSkillsNetwork20297740-2021-01-01#plotting)<br>
*   [Vẽ biểu đồ với Dataframe](http://pandas.pydata.org/pandas-docs/stable/api.html?utm_medium=Exinfluencer\&utm_source=Exinfluencer\&utm_content=000026UJ\&utm_term=10006555\&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDV0101ENSkillsNetwork20297740-2021-01-01#api-dataframe-plotting)


# Biểu đồ đường (Series/Dataframe) <a id="12"></a>


**Biểu đồ đường là gì và tại sao chúng ta sử dụng nó?**

Biểu đồ đường là một loại biểu đồ hiển thị thông tin dưới dạng một chuỗi các điểm dữ liệu được gọi là 'marker' (điểm đánh dấu) được kết nối bằng các đoạn thẳng. Đây là loại biểu đồ cơ bản, phổ biến trong nhiều lĩnh vực.
Chúng ta sử dụng biểu đồ đường khi có tập dữ liệu liên tục. Đây là cách phù hợp nhất cho việc trực quan hóa dữ liệu theo xu hướng trong một khoảng thời gian.


**Hãy bắt đầu với case study sau:**

Năm 2010, Haiti phải hứng chịu thảm họa động đất mạnh 7,0 độ richter. Trận động đất này đã gây ra sự tàn phá và thiệt hại về người trên diện rộng, khoảng 3 triệu người đã bị ảnh hưởng bởi thảm họa thiên nhiên này. Chính phủ Canada đã tăng cường nỗ lực tiếp nhận người tị nạn từ Haiti. Chúng ta có thể nhanh chóng hình dung nỗ lực này bằng cách sử dụng biểu đồ `Đường`:

**Câu hỏi:** Vẽ biểu đồ đường biểu thị di cư từ Haiti, sử dụng `df.plot()`.


Đầu tiên, chúng ta sẽ trích xuất chuỗi dữ liệu cho Haiti.


In [ ]:
haiti = df_can.loc['Haiti', years] # nhập các năm 1980 - 2013 để loại trừ cột 'total'
haiti.head()

NameError: name 'years' is not defined

Tiếp theo, chúng ta sẽ vẽ một biểu đồ đường bằng cách thêm `.plot ()` vào dataframe `haiti`.


In [ ]:
haiti.plot()

_pandas_ tự động phân bố trục x với các giá trị chỉ số (năm) và trục y với các giá trị cột (dân số). Tuy nhiên, hãy lưu ý cách các năm không được hiển thị vì chúng thuộc kiểu _string_. Do đó, hãy thay đổi kiểu giá trị chỉ mục thành _integer_ để vẽ biểu đồ.

Ngoài ra, hãy gắn nhãn trục x và y bằng `plt.title()`, `plt.ylabel()`, và `plt.xlabel()` như sau:


In [ ]:
haiti.index = haiti.index.map(int) # hãy thay đổi giá trị chỉ số của Hati thành kiểu integer để vẽ biểu đồ
haiti.plot(kind='line')

plt.title('Immigration from Haiti')
plt.ylabel('Number of immigrants')
plt.xlabel('Years')

plt.show() # cần dòng này để thể hiện các cập nhật cho hình

Chúng ta có thể nhận thấy rõ ràng số lượng người nhập cư từ Haiti đã tăng vọt như thế nào từ năm 2010 khi Canada tăng cường nỗ lực tiếp nhận người tị nạn từ Haiti. Hãy chú thích mức tăng đột biến này trong biểu đồ bằng phương thức `plt.text ()`.


In [ ]:
haiti.plot(kind='line')

plt.title('Immigration from Haiti')
plt.ylabel('Number of Immigrants')
plt.xlabel('Years')

# chú thích động đất năm 2010.
# cú pháp: plt.text(x, y, label)
plt.text(2000, 6000, '2010 Earthquake') # xem ghi chú bên dưới

plt.show()

Chỉ với một vài dòng code, bạn đã có thể nhanh chóng xác định và hình dung sự gia tăng đột biến trong lượng nhập cư!

Ghi chú vắn tắt về gía trị x và y trong `plt.text(x, y, label)`:

```
 Vì trục x (năm) là kiểu 'integer', chúng ta chỉ định x là năm. Trục y (số người nhập cư) là kiểu 'integer', vì vậy chúng ta có thể chỉ định giá trị y = 6000.
```

```python
    plt.text(2000, 6000, '2010 Earthquake') # các năm được lưu như kiểu int
```

```
Nếu năm được lưu trữ dưới dạng 'string', chúng ta cần chỉ định x là vị trí chỉ mục của năm. Ví dụ: chỉ số thứ 20 là năm 2000 vì nó là năm thứ 20 với năm cơ sở là 1980.
```

```python
    plt.text(20, 6000, '2010 Earthquake') # các năm được lưu như kiểu int
```

```
Chúng ta sẽ tìm hiểu các phương pháp chú thích nâng cao trong các mô-đun sau.
```


Chúng ta có thể dễ dàng thêm nhiều quốc gia hơn vào biểu đồ để khiến các so sánh về nhập cư từ các quốc gia khác nhau có ý nghĩa hơn.

**Câu hỏi:** Hãy so sánh số lượng người nhập cư từ Ấn Độ và Trung Quốc từ năm 1980 đến năm 2013.


Bước 1: Lấy tập dữ liệu cho Trung Quốc và Ấn Độ, đồng thời hiển thị dataframe.


In [ ]:
### nhập đáp án của bạn tại đây



<details><summary>Click vào đây để xem lời giải mẫu trong python</summary>

```python
    #The correct answer is:
    df_CI = df_can.loc[['India', 'China'], years]
    df_CI.head()
```

</details>


Bước 2: Vẽ đồ thị. Chúng ta sẽ chỉ định rõ ràng biểu đồ đường bằng cách truyền tham số `kind` cho `plot()`.


In [ ]:
### nhập đáp án của bạn tại đây



<details><summary>Click vào đây để xem lời giải mẫu trong python</summary>

```python
    #The correct answer is:
    df_CI.plot(kind='line')
```

</details>


Có vẻ không đúng...

Hãy nhớ lại rằng _pandas_ vẽ các chỉ số trên trục x và các cột dưới dạng các đường riêng lẻ trên trục y. Vì `df_CI` là một dataframe, `country` là chỉ số và `years` là các cột, nên trước tiên chúng ta phải hoán vị dataframe bằng phương thức `transpose()`để hoán đổi hàng và cột.


In [ ]:
df_CI = df_CI.transpose()
df_CI.head()

*pandas* sẽ tự động minh họa hai quốc gia trên cùng một biểu đồ. Hãy tiếp tục và vẽ dataframe chuyển vị mới. Hãy đảm bảo thêm tiêu đề cho biểu đồ và gắn nhãn cho các trục.


In [ ]:
### nhập đáp án của bạn tại đây





<details><summary>Click vào đây để xem lời giải mẫu trong python</summary>

```python
    #The correct answer is:
    df_CI.index = df_CI.index.map(int) # hãy thay đổi giá trị chỉ mục của df_CI thành kiểu integer để vẽ biểu đồ
    df_CI.plot(kind='line')

    plt.title('Immigrants from China and India')
    plt.ylabel('Number of Immigrants')
    plt.xlabel('Years')

    plt.show()
```

</details>


<br>Từ sơ đồ trên, chúng ta có thể thấy rằng Trung Quốc và Ấn Độ có xu hướng nhập cư rất giống nhau qua các năm.


_Ghi chú_: Tại sao chúng ta không cần chuyển dataframe của Haiti trước khi vẽ biểu đồ (như đã làm với df_CI)?

Đó là do `haiti` là một chuỗi ngược với dataframe và có chỉ số là các năm như được hiển thị bên dưới.

```python
print(type(haiti))
print(haiti.head(5))
```

> class 'pandas.core.series.Series' <br>
> 1980    1666 <br>
> 1981    3692 <br>
> 1982    3498 <br>
> 1983    2860 <br>
> 1984    1418 <br>
> Name: Haiti, dtype: int64 <br>


Biểu đồ đường là một công cụ tiện dụng để hiển thị một số biến phụ thuộc so với một biến độc lập. Tuy nhiên, chúng tôi khuyến nghị không nên có quá 5-10 đường trên một biểu đồ; nhiều hơn thế sẽ khó phân tích.


**Câu hỏi:** So sánh xu hướng của 5 quốc gia có nhiều người nhập cư vào Canada nhất.


In [ ]:
### nhập đáp án của bạn tại đây





<details><summary>Click vào đây để xem lời giải mẫu trong python</summary>

```python
    #The correct answer is:    
    #Bước 1: Nhận tập dữ liệu. Nhớ lại rằng chúng ta đã tạo cột Total để tính toán lượng nhập cư tích lũy theo quốc gia.
    #Chúng ta sẽ sắp xếp trên cột này để lấy top 5 quốc gia bằng phương thức pandas sort_values().
    
    inplace = True paramemter saves the changes to the original df_can dataframe
    df_can.sort_values(by='Total', ascending=False, axis=0, inplace=True)

    # lấy 5 mục đầu
    df_top5 = df_can.head(5)

    # chuyển đổi dataframe
    df_top5 = df_top5[years].transpose()

    print(df_top5)


    #Bước 2: Vẽ dataframe. Để giúp biểu đồ dễ đọc hơn, chúng ta sẽ thay đổi kích thước bằng tham số `figsize`.
    df_top5.index = df_top5.index.map(int) # hãy thay đổi các giá trị chỉ mục của df_top5 thành kiểu integer để vẽ biểu đồ
    df_top5.plot(kind='line', figsize=(14, 8)) # truyền một tuple (x, y) size



    plt.title('Immigration Trend of Top 5 Countries')
    plt.ylabel('Number of Immigrants')
    plt.xlabel('Years')


    plt.show()

```

</details>


### Các biểu đồ khác

Chúc mừng bạn đã học cách làm sạch dữ liệu với python và tạo biểu đồ đường bằng Matplotlib. Ngoài biểu đồ đường mặc định,còn nhiều phong cách vẽ biểu đồ khác, tất cả đều có thể được truy cập bằng cách chuyển từ khóa `kind` sang `plot()`. Danh sách đầy đủ các biểu đồ có sẵn như sau:

*   `bar` cho biểu đồ cột
*   `barh` cho biểu đồ cột ngang
*   `hist` cho biểu đồ tần suất
*   `box` cho biểu đồ hộp
*   `kde` hoặc `density` cho biểu đồ mật độ
*   `area` cho biểu đồ vùng
*   `pie` cho biểu đồ tròn
*   `scatter` cho biểu đồ phân tán
*   `hexbin` cho biểu đồ hexbin
